File: frontend/lib/binomial.ts
export function crrPrice(
  S: number, K: number, T: number, r: number, sigma: number,
  optionType: "call" | "put",
  american: boolean = false,
  N: number = 200,
): number {
  const dt = T / N;
  const u = Math.exp(sigma * Math.sqrt(dt));
  const d = 1 / u;
  const disc = Math.exp(-r * dt);
  const p = (Math.exp(r * dt) - d) / (u - d);
  const q = 1 - p;

  // Terminal payoffs
  const V = new Float64Array(N + 1);
  for (let j = 0; j <= N; j++) {
    const ST = S * Math.pow(u, j) * Math.pow(d, N - j);
    V[j] = optionType === "call" ? Math.max(ST - K, 0) : Math.max(K - ST, 0);
  }

  // Backward induction
  for (let i = N - 1; i >= 0; i--) {
    for (let j = 0; j <= i; j++) {
      V[j] = disc * (p * V[j + 1] + q * V[j]);
      if (american) {
        const Snode = S * Math.pow(u, j) * Math.pow(d, i - j);
        const intrinsic = optionType === "call" ? Math.max(Snode - K, 0) : Math.max(K - Snode, 0);
        V[j] = Math.max(V[j], intrinsic);
      }
    }
  }

  return V[0];
}

File: src/binomial.py
"""Binomial tree pricer (Cox-Ross-Rubinstein) — European and American options."""

import math
import numpy as np


def crr_price(
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    option_type: str = "call",
    american: bool = False,
    N: int = 200,
) -> float:
    """
    Price a European or American option using the CRR binomial tree.

    Parameters
    ----------
    S           : spot price
    K           : strike price
    T           : time to maturity in years
    r           : continuous risk-free rate
    sigma       : annualised volatility
    option_type : "call" or "put"
    american    : True for American-style early exercise
    N           : number of time steps (more steps = more accurate)
    """
    if option_type not in ("call", "put"):
        raise ValueError(f"option_type must be 'call' or 'put', got '{option_type}'")

    dt = T / N

    # CRR parameters
    u = math.exp(sigma * math.sqrt(dt))
    d = 1.0 / u
    disc = math.exp(-r * dt)
    p = (math.exp(r * dt) - d) / (u - d)   # risk-neutral probability
    q = 1.0 - p

    # ── Pass 1: terminal asset prices and payoffs ──
    # j=0 → all down, j=N → all up
    j = np.arange(N + 1)
    S_T = S * (u ** j) * (d ** (N - j))

    if option_type == "call":
        V = np.maximum(S_T - K, 0.0)
    else:
        V = np.maximum(K - S_T, 0.0)

    # ── Pass 2: backward induction ──
    for i in range(N - 1, -1, -1):
        # Roll back one step: V[j] comes from V[j+1] (up) and V[j] (down)
        V = disc * (p * V[1:] + q * V[:-1])

        if american:
            # Asset prices at this earlier node
            j = np.arange(i + 1)
            S_node = S * (u ** j) * (d ** (i - j))
            intrinsic = np.maximum(S_node - K, 0.0) if option_type == "call" else np.maximum(K - S_node, 0.0)
            V = np.maximum(V, intrinsic)

    return float(V[0])
